File: ios/DoMemory/DoMemory/Services/Levels/LevelCurve.swift
//
//  LevelCurve.swift
//  DoMemory
//
//  Pure difficulty curve for Levels mode: how many pairs and how much time a
//  given level number gets. Deliberately table-driven (anchor points +
//  linear interpolation) rather than a closed-form formula, since the
//  approved curve is sub-linear and this keeps it easy to read and retune.
//

import Foundation

enum LevelCurve {
    /// (level, pairs) anchors. Value is held constant past the last anchor.
    private static let pairAnchors: [(level: Int, value: Int)] = [
        (1, 3), (5, 4), (10, 6), (25, 9), (50, 12)
    ]

    /// (level, seconds) anchors. Value is held constant past the last anchor.
    private static let secondAnchors: [(level: Int, value: Int)] = [
        (1, 90), (5, 85), (10, 75), (25, 60), (50, 45), (80, 35)
    ]

    /// (level, mistakes) anchors — the budget of failed matches before the
    /// level is lost. Tracks roughly `pairs + 2`: even perfect recall costs
    /// about N/2–N mismatches on an N-pair board, so a budget at or below the
    /// pair count would be effectively unwinnable.
    private static let failureAnchors: [(level: Int, value: Int)] = [
        (1, 4), (5, 6), (10, 8), (25, 11), (50, 14)
    ]

    /// Number of pairs of cards for a level. Non-decreasing as level increases; capped at 12.
    static func pairs(for level: Int) -> Int {
        interpolate(level: level, anchors: pairAnchors)
    }

    /// Time limit in seconds for a level. Non-increasing as level increases; floored at 35.
    static func seconds(for level: Int) -> Int {
        interpolate(level: level, anchors: secondAnchors)
    }

    /// How many failed matches a level allows before it's lost. Always greater
    /// than `pairs(for:)` so every level stays winnable; capped at 14.
    static func maxFailures(for level: Int) -> Int {
        interpolate(level: level, anchors: failureAnchors)
    }

    private static func interpolate(level: Int, anchors: [(level: Int, value: Int)]) -> Int {
        let clampedLevel = max(1, level)

        guard let first = anchors.first, let last = anchors.last else { return 0 }
        if clampedLevel <= first.level { return first.value }
        if clampedLevel >= last.level { return last.value }

        for index in 1..<anchors.count {
            let lo = anchors[index - 1]
            let hi = anchors[index]
            guard clampedLevel <= hi.level else { continue }

            let span = Double(hi.level - lo.level)
            let progress = Double(clampedLevel - lo.level) / span
            let value = Double(lo.value) + progress * Double(hi.value - lo.value)
            return Int(value.rounded())
        }
        return last.value
    }
}

File: android/app/src/main/java/com/ezequielbrrt/domemory/services/levels/LevelCurve.kt
package com.ezequielbrrt.domemory.services.levels

import kotlin.math.roundToInt

/**
 * The endless-Levels difficulty curve (spec 7.1). Table-driven with linear
 * interpolation between anchors; the last anchor's value is held forever after.
 *
 * Deliberately not a formula — the curve is sub-linear and this keeps it retunable.
 *
 * `maxFailures` tracks roughly `pairs + 2` on purpose: even perfect recall costs
 * N/2–N mismatches on an N-pair board, so a budget at or below the pair count would
 * make levels unwinnable.
 */
object LevelCurve {

    private data class Anchor(val level: Int, val value: Int)

    private val pairAnchors = listOf(
        Anchor(1, 3), Anchor(5, 4), Anchor(10, 6), Anchor(25, 9), Anchor(50, 12),
    )
    private val secondAnchors = listOf(
        Anchor(1, 90), Anchor(5, 85), Anchor(10, 75),
        Anchor(25, 60), Anchor(50, 45), Anchor(80, 35),
    )
    private val failureAnchors = listOf(
        Anchor(1, 4), Anchor(5, 6), Anchor(10, 8), Anchor(25, 11), Anchor(50, 14),
    )

    fun pairs(level: Int): Int = interpolate(level, pairAnchors)

    fun seconds(level: Int): Int = interpolate(level, secondAnchors)

    fun maxFailures(level: Int): Int = interpolate(level, failureAnchors)

    /**
     * The pair cap, held forever past the last anchor. This is also the floor on a
     * season's emoji pool (spec 9.3): a season with fewer distinct emoji than this
     * cannot fill its own level-25+ boards. Derive that constant from here so
     * retuning the curve moves the floor with it.
     */
    val maxPairs: Int get() = pairAnchors.last().value

    /** The pie indicator appears from this level on, in Levels and Seasons (spec 3.4). */
    const val PIE_FROM_LEVEL = 25

    fun showsPie(level: Int): Boolean = level >= PIE_FROM_LEVEL

    private fun interpolate(level: Int, anchors: List<Anchor>): Int {
        val l = maxOf(1, level)
        val first = anchors.first()
        val last = anchors.last()
        if (l <= first.level) return first.value
        if (l >= last.level) return last.value

        val hiIndex = anchors.indexOfFirst { l <= it.level }
        val hi = anchors[hiIndex]
        val lo = anchors[hiIndex - 1]
        val t = (l - lo.level).toDouble() / (hi.level - lo.level).toDouble()
        return (lo.value + t * (hi.value - lo.value)).roundToInt()
    }
}
